# Lasso & Ridge Regression:

![](https://tyami.github.io/assets/images/post/ML/2020-10-15-ridge-and-lasso/2020-10-15-ridge-and-lasso-1-overview.png)

## Outline:
* What is Lasso & Ridge Regression?
* Practical

## What is Lasso & Ridge Regression?

<img src='https://miro.medium.com/max/552/1*zuyjF8wKPxCC7QSoidpwxA.png' width='600' hight='600'>

* Before learning L&R Regression you need to understand the Linear Regression.
* In above digram you can see the three lines for different regressions.
* **Linear Regression:** Try to best fit line on the dataset. But in this process our model may be overfit and during the test it gives us maximum error.
* To reduce this error we need to penalized our model.
* To penalize our model we use Regularized methods like: `L_1 & L_2`.
* Using **Lasso & Ridge techniques** we generalized our Linear Regression Model. So that during model deployment we could reduce or clear the maximum errors.

## Mathematical Expression Of Lasso & Ridge Regression:
![](https://static.wixstatic.com/media/b5b5ca_cb3522fa2f8446498f164a2e694efbd2~mv2.png/v1/fill/w_591,h_436,al_c,lg_1,q_85,enc_auto/b5b5ca_cb3522fa2f8446498f164a2e694efbd2~mv2.png)

## Ridge Regression:
![](https://static.wixstatic.com/media/b5b5ca_2f72ca6ab0ea4792a9fbd35230bb9ef7~mv2.png/v1/fill/w_740,h_91,al_c,q_85,usm_0.66_1.00_0.01,enc_auto/b5b5ca_2f72ca6ab0ea4792a9fbd35230bb9ef7~mv2.png)

### Note:
* The LHS side of the Ridge Regression equation is called `Cos function`.
* **Cos Function:** Square of the difference of the actual and predicted value.
* We try to minimize our `cos function` until `zero`. 
* To generalize our Linear Regression model add `lambda` and multiply with the **square of the `slop`**.

## Lasso Regression
![](https://static.wixstatic.com/media/b5b5ca_7c09a96763784778ae7ddcf0aa4dde88~mv2.png/v1/fill/w_740,h_85,al_c,q_85,usm_0.66_1.00_0.01,enc_auto/b5b5ca_7c09a96763784778ae7ddcf0aa4dde88~mv2.png)

### Note:
* The LHS side of the Lasso Regression equation is called `Cos function`.
* **Cos Function:** Square of the difference of the actual and predicted value.
* We try to minimize our `cos function` until `zero`. 
* To generalize our Linear Regression model add `lambda` and multiply with the **magnitude of the `slop`**.

## Practical: We use the Banglore House Prediction cleaned data

### Step 1: Loading the libraries:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Step 2: Loading The Dataset:

In [2]:
df=pd.read_csv('Cleaned Banglore House Price Prediction Data.csv')

In [3]:
df.head()

area_type  availability  location  bath  balcony   price  new_size  \
0          3            38       341   2.0      1.0   39.07         2   
1          2            77       251   5.0      3.0  120.00         4   
2          0            77       964   2.0      3.0   62.00         3   
3          3            77       629   3.0      1.0   95.00         3   
4          3            77       592   2.0      1.0   51.00         2   

   new_total_sqft  price_per_sqft  
0          1056.0     3699.810606  
1          2600.0     4615.384615  
2          1440.0     4305.555556  
3          1521.0     6245.890861  
4          1200.0     4250.000000

#### Note:
* This is our cleaned data we have cleaned this in Linera Regression Model.
* Now again we use `Linear Regression` use perform the `Ridge & Lasso` techniques.

### Step 3: Performing The Linear Regression Method:

In [4]:
# Spliting the dataset into Target (y) and Features(X) variables:
X=df.drop('price',axis=1)
y=df['price']

print('Shape of X:', X.shape)
print('Shape of y:',y.shape)

Shape of X: (11018, 8)
Shape of y: (11018,)


In [5]:
# Importing the `train_test_split` library form sklearn.model_selection
from sklearn.model_selection import train_test_split

In [6]:
# Spliting the data into train & test part:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=.2, random_state=6)

print('Shape of X_train = ', X_train.shape)
print('Shape of y_train = ', y_train.shape)
print('Shape of X_test = ', X_test.shape)
print('Shape of y_test = ', y_test.shape)

Shape of X_train =  (8814, 8)
Shape of y_train =  (8814,)
Shape of X_test =  (2204, 8)
Shape of y_test =  (2204,)


In [7]:
'''
Feature scalling of the dataset: We'll normalize the different numerica value of the data 
                                 so that machine can easly understand.
'''
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
sc.fit(X_train)
X_train=sc.transform(X_train)
X_test=sc.transform(X_test)

In [8]:
# Linear Regression ML Model Training:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()

lr.fit(X_train,y_train)

LinearRegression()

In [9]:
'''Exploring The Trained Data
'''
# Seeing the coefficent values:
lr.coef_

array([ 0.15192576, -0.42192927, -0.22437487, -0.17351644,  0.15535644,
       -4.40445112, 55.26726976, 25.18508851])

In [10]:
# Seeing the intersect value:
lr.intercept_

82.08727989562063

In [11]:
# Testing the model:
lr.predict(X_test)

array([107.6745738 ,  34.500194  ,  60.94472328, ...,  82.4907132 ,
       147.67319159, 163.35794178])

In [12]:
# Verifing the actual values:
y_test

4064     104.00
7855      42.81
10856     61.11
4491      47.00
9184      33.50
          ...  
3991      75.00
26        57.39
9239      80.00
8944     150.00
10194    170.00
Name: price, Length: 2204, dtype: float64

In [24]:
# Checking the accuracy of the model:
lr.score(X_test,y_test)

0.94471512931249

### Conclusion:
* Our model is give us `94.4715%` accureacy which is impressive.
* But we need to improve the accuracy.

### Step 4: Implementing Ridge & Lasso Regression:

In [14]:
# importing the library:
from sklearn.linear_model import Ridge, Lasso

In [15]:
# 1. Ridge Regression:implimenting onto the train & test dataset:
rd=Ridge()

# Fitting onto the dataset
rd.fit(X_train,y_train)

# Calculating the score
rd.score(X_test,y_test)

0.9447160002376536

### Conclusion:
* Our model is give us `94.4716%` accureacy which is incremant of just `.0001%`.
* But we need to improve the accuracy.

In [16]:
# 2. Lasso Regression: implimenting onto the trian & test dataset:
ls=Lasso()

# Fitting the dataset:
ls.fit(X_train, y_train)

# Calculating the score
ls.score(X_test,y_test)

0.9433947683102221

### Conclusion:
* Our model is give us `94.3394%` accureacy which is decremant of `0.1321%`.
* But we need to improve the accuracy.

### Step 5: Let's change the `alpha`:

In [20]:
# 1.1: Ridge Regression:implimenting onto the train & test dataset:
rd2=Ridge(alpha=3)

# Fitting onto the dataset
rd2.fit(X_train,y_train)

# Calculating the score
rd2.score(X_test,y_test)

0.9447176552945601

### Conclusion:
* Nothing changed.

In [23]:
# 2.1: Lasso Regression: implimenting onto the trian & test dataset:
ls2=Lasso(alpha=2)

# Fitting the dataset:
ls2.fit(X_train, y_train)

# Calculating the score
ls2.score(X_test,y_test)

0.9408991437885685

### Conclusion:
* Decreased our acurracy.

## Remark:
* As we are changing the value of `alpha` our model's accuracy is increasing or decreasing slowly.
* So to automate this process we use **Hyper Parameter Tunning Method** so that our model predict more accurate value.